### PR TITLE
[data] New executor [5/n]--- Add legacy compatibility shims

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

After this PR, you can try using the new bulk execution backend by setting `RAY_DATASET_NEW_EXECUTION_BACKEND=1`. Note that a number of features are not yet implemented, such as actor pools, callable classes--- see https://github.com/ray-project/ray/pull/30903 for full details.

The plan moving forward from here is:
1. Pass all existing unit tests and nightlies with the new backend, pushing bug fixes and features into master as separate PRs.
2. Flip the flag to enable the new backend in a separate PR.

This PR only contains a minimal "sanity" test to ensure the new backend isn't completely broken by changes.

### CODE REVIEW

**Comment by stephanie-wang** [comment]
The legacy integration looks good but I'm confused by the block freeing parts. Maybe it's just because I'm missing context, but can you explain why we need the new calls to `trace_allocation`?

**Comment by ericl** [comment]
Here's a brief overview of the state of block freeing:

In the legacy implementation, there are two concepts to be aware of: (1) "should clear blocks"--- this is a flag that tells the plan.py file whether it should be able to clear input blocks, and (2) "owned by consumer"--- this is a flag that tells iterators whether they can free blocks as they iterate over them.

In the new execution backend, there is only a single "owned" concept for a BundleRef--- if it's owned, it is always safe to delete. The operator implementations are responsible for calling bundle.destroy_if_owned() when they no longer need bundles.

In legacy_compat.py we bridge this gap by roughly translating "should clear blocks" -> "owned", and only setting "owned" for output blocks if the plan is "run by consumer". This is probably the most confusing part.

Re: the trace allocation changes, those are only to enable better debugging (https://github.com/ray-project/ray/pull/31283).

File: python/ray/data/_internal/block_batching.py [hunk 1]
```diff
@@ -183,8 +184,10 @@ def _sliding_window(iterable: Iterable, n: int, clear_block_after_read: bool = F
         yield tuple(window)
     for elem in it:
         block_ref = window.popleft()
-        if clear_block_after_read:
-            ray._private.internal_api.free(block_ref, local_only=False)
+        if clear_block_after_read and DatasetContext.get_current().eager_free:
+            trace_deallocation(block_ref, "block_batching._sliding_window")
```
Comment (jianoaix) [inline python/ray/data/_internal/block_batching.py:188]: Since this tracing utility is for debugging only, why is this also used on prod path?

Comment (ericl) [reply python/ray/data/_internal/block_batching.py:188]: The call is always there, but it's a no-op when the flag is disabled.

File: python/ray/data/_internal/plan.py [hunk 2]
```diff
@@ -328,33 +328,62 @@ def execute(
...
-                logger.get_logger(log_to_stdout=context.enable_auto_log_stats).info(
-                    stats_summary_string,
+                executor = BulkExecutor(ExecutionOptions())
+                blocks = execute_to_legacy_block_list(
+                    executor,
+                    self,
+                    allow_clear_input_blocks=allow_clear_input_blocks,
+                    dataset_uuid=self._dataset_uuid,
                 )
+                # TODO(ekl) we shouldn't need to set this in the future once we move
```
Comment (c21) [inline python/ray/data/_internal/plan.py:352]: can we file a github issue for this TODO?

**Review by c21 — APPROVED** [review]
LGTM

File: python/ray/data/_internal/plan.py (hunk 2 above)
Comment (ericl) [reply python/ray/data/_internal/plan.py:352]: I'll do this after the REP is approved.

File: python/ray/data/_internal/execution/legacy_compat.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        raise NotImplementedError
+
+
+def _bundles_to_block_list(bundles: Iterator[RefBundle]) -> BlockList:
+    blocks, metadata = [], []
+    for ref_bundle in bundles:
+        for block, meta in ref_bundle.blocks:
+            blocks.append(block)
+            metadata.append(meta)
+    owns_blocks = all(b.owns_blocks for b in bundles)
```
Comment (jianoaix) [inline python/ray/data/_internal/execution/legacy_compat.py:198]: I think this is logically correct with a conjunction, I wonder in what cases we will have different values of b.owns_blocks?

**Review by jianoaix — COMMENTED** [review]
LGTM

File: python/ray/data/_internal/execution/legacy_compat.py (hunk 3 above)
Comment (ericl) [reply python/ray/data/_internal/execution/legacy_compat.py:198]: I think it could happen theoretically with a union stage.

**Review by clarkzinzow — APPROVED** [review]
LGTM, actually ended up being a lot simpler than expected!

**Pull request merged**